File: src/main/ioke/lang/Pair.java
/*
 * See LICENSE file in distribution for copyright and licensing information.
 */
package ioke.lang;

import java.util.List;
import java.util.Map;

import ioke.lang.exceptions.ControlFlow;

/**
 *
 * @author <a href="mailto:ola.bini@gmail.com">Ola Bini</a>
 */
public class Pair extends IokeData {
    private Object first;
    private Object second;

    public Pair(Object first, Object second) {
        this.first = first;
        this.second = second;
    }

    public static Object getFirst(Object pair) {
        return ((Pair)IokeObject.data(pair)).getFirst();
    }

    public static Object getSecond(Object pair) {
        return ((Pair)IokeObject.data(pair)).getSecond();
    }

    public Object getFirst() {
        return first;
    }

    public Object getSecond() {
        return second;
    }

    @Override
    public void init(IokeObject obj) throws ControlFlow {
        final Runtime runtime = obj.runtime;

        obj.setKind("Pair");
        obj.mimics(IokeObject.as(runtime.mixins.getCell(null, null, "Enumerable"), null), runtime.nul, runtime.nul);
        obj.mimics(IokeObject.as(runtime.mixins.getCell(null, null, "Comparing"), null), runtime.nul, runtime.nul);

        obj.registerMethod(runtime.newJavaMethod("Returns the first value", new TypeCheckingJavaMethod.WithNoArguments("first", runtime.pair) {
                @Override
                public Object activate(IokeObject method, Object on, List<Object> args, Map<String, Object> keywords, IokeObject context, IokeObject message) throws ControlFlow {
                    return ((Pair)IokeObject.data(on)).first;
                }
            }));

        obj.registerMethod(runtime.newJavaMethod("Returns the first value", new TypeCheckingJavaMethod.WithNoArguments("key", runtime.pair) {
                @Override
                public Object activate(IokeObject method, Object on, List<Object> args, Map<String, Object> keywords, IokeObject context, IokeObject message) throws ControlFlow {
                    return ((Pair)IokeObject.data(on)).first;
                }
            }));

        obj.registerMethod(runtime.newJavaMethod("Returns the second value", new TypeCheckingJavaMethod.WithNoArguments("second", runtime.pair) {
                @Override
                public Object activate(IokeObject method, Object on, List<Object> args, Map<String, Object> keywords, IokeObject context, IokeObject message) throws ControlFlow {
                    return ((Pair)IokeObject.data(on)).second;
                }
            }));

        obj.registerMethod(runtime.newJavaMethod("Returns the second value", new TypeCheckingJavaMethod.WithNoArguments("value", runtime.pair) {
                @Override
                public Object activate(IokeObject method, Object on, List<Object> args, Map<String, Object> keywords, IokeObject context, IokeObject message) throws ControlFlow {
                    return ((Pair)IokeObject.data(on)).second;
                }
            }));

        obj.registerMethod(runtime.newJavaMethod("Returns a text inspection of the object", new TypeCheckingJavaMethod.WithNoArguments("inspect", runtime.pair) {
                @Override
                public Object activate(IokeObject method, Object on, List<Object> args, Map<String, Object> keywords, IokeObject context, IokeObject message) throws ControlFlow {
                    return method.runtime.newText(Pair.getInspect(on));
                }
            }));

        obj.registerMethod(runtime.newJavaMethod("Returns a brief text inspection of the object", new TypeCheckingJavaMethod.WithNoArguments("notice", runtime.pair) {
                @Override
                public Object activate(IokeObject method, Object on, List<Object> args, Map<String, Object> keywords, IokeObject context, IokeObject message) throws ControlFlow {
                    return method.runtime.newText(Pair.getNotice(on));
                }
            }));
    }

    public IokeData cloneData(IokeObject obj, IokeObject m, IokeObject context) {
        return new Pair(first, second);
    }


    public static String getInspect(Object on) throws ControlFlow {
        return ((Pair)(IokeObject.data(on))).inspect(on);
    }

    public static String getNotice(Object on) throws ControlFlow {
        return ((Pair)(IokeObject.data(on))).notice(on);
    }

    @Override
    public boolean isEqualTo(IokeObject self, Object other) {
        return ((other instanceof IokeObject) && 
                (IokeObject.data(other) instanceof Pair) 
                && this.first.equals(((Pair)IokeObject.data(other)).first)
                && this.second.equals(((Pair)IokeObject.data(other)).second));
    }

    @Override
    public String toString() {
        return "" + first + " => " + second;
    }

    @Override
    public String toString(IokeObject obj) {
        return "" + first + " => " + second;
    }

    public String inspect(Object obj) throws ControlFlow {
        StringBuilder sb = new StringBuilder();

        sb.append(IokeObject.inspect(first));
        sb.append(" => ");
        sb.append(IokeObject.inspect(second));

        return sb.toString();
    }

    public String notice(Object obj) throws ControlFlow {
        StringBuilder sb = new StringBuilder();

        sb.append(IokeObject.notice(first));
        sb.append(" => ");
        sb.append(IokeObject.notice(second));

        return sb.toString();
    }
}// Pair

File: src/ikc/main/Ioke.Lang/Pair.cs
namespace Ioke.Lang {
    using System.Text;

    public class Pair : IokeData {
        object first;
        object second;

        public Pair(object first, object second) {
            this.first = first;
            this.second = second;
        }

        public static object GetFirst(object pair) {
            return ((Pair)IokeObject.dataOf(pair)).First;
        }

        public static object GetSecond(object pair) {
            return ((Pair)IokeObject.dataOf(pair)).Second;
        }

        public object First {
            get { return first; }
        }

        public object Second {
            get { return second; }
        }

        public override void Init(IokeObject obj) {
            Runtime runtime = obj.runtime;

            obj.Kind = "Pair";
            obj.Mimics(IokeObject.As(IokeObject.FindCell(runtime.Mixins, "Enumerable"), null), runtime.nul, runtime.nul);
            obj.Mimics(IokeObject.As(IokeObject.FindCell(runtime.Mixins, "Comparing"), null), runtime.nul, runtime.nul);

            obj.RegisterMethod(obj.runtime.NewNativeMethod("returns a hash for the pair",
                                                           new NativeMethod.WithNoArguments("hash", (method, context, message, on, outer) => {
                                                                   outer.ArgumentsDefinition.CheckArgumentCount(context, message, on);
                                                                   var one = ((Pair)IokeObject.dataOf(on)).first.GetHashCode();
                                                                   var two = ((Pair)IokeObject.dataOf(on)).second.GetHashCode();
                                                                   return context.runtime.NewNumber(one + 13*two);
                                                               })));

            obj.RegisterMethod(runtime.NewNativeMethod("returns true if the left hand side pair is equal to the right hand side pair.",
                                                       new TypeCheckingNativeMethod("==", TypeCheckingArgumentsDefinition.builder()
                                                                                    .ReceiverMustMimic(runtime.Pair)
                                                                                    .WithRequiredPositional("other")
                                                                                    .Arguments,
                                                                                    (method, on, args, keywords, context, message) => {
                                                                                        Pair d = (Pair)IokeObject.dataOf(on);
                                                                                        object other = args[0];

                                                                                        return ((other is IokeObject) &&
                                                                                                (IokeObject.dataOf(other) is Pair)
                                                                                                && d.first.Equals(((Pair)IokeObject.dataOf(other)).first)
                                                                                                && d.second.Equals(((Pair)IokeObject.dataOf(other)).second)) ? context.runtime.True : context.runtime.False;
                                                                                    })));

            obj.RegisterMethod(runtime.NewNativeMethod("Returns the first value",
                                                       new TypeCheckingNativeMethod.WithNoArguments("first", obj,
                                                                                                    (method, on, args, keywords, context, message) => {
                                                                                                        return ((Pair)IokeObject.dataOf(on)).first;
                                                                                                    })));

            obj.RegisterMethod(runtime.NewNativeMethod("Returns the first value",
                                                       new TypeCheckingNativeMethod.WithNoArguments("key", obj,
                                                                                                    (method, on, args, keywords, context, message) => {
                                                                                                        return ((Pair)IokeObject.dataOf(on)).first;
                                                                                                    })));


            obj.RegisterMethod(runtime.NewNativeMethod("Returns the second value",
                                                       new TypeCheckingNativeMethod.WithNoArguments("second", obj,
                                                                                                    (method, on, args, keywords, context, message) => {
                                                                                                        return ((Pair)IokeObject.dataOf(on)).second;
                                                                                                    })));

            obj.RegisterMethod(runtime.NewNativeMethod("Returns the second value",
                                                       new TypeCheckingNativeMethod.WithNoArguments("value", obj,
                                                                                                    (method, on, args, keywords, context, message) => {
                                                                                                        return ((Pair)IokeObject.dataOf(on)).second;
                                                                                                    })));

            obj.RegisterMethod(runtime.NewNativeMethod("Returns a text inspection of the object",
                                                       new TypeCheckingNativeMethod.WithNoArguments("inspect", obj,
                                                                                                    (method, on, args, keywords, context, message) => {
                                                                                                        return method.runtime.NewText(Pair.GetInspect(on));
                                                                                                    })));

            obj.RegisterMethod(runtime.NewNativeMethod("Returns a brief text inspection of the object",
                                                       new TypeCheckingNativeMethod.WithNoArguments("notice", obj,
                                                                                                    (method, on, args, keywords, context, message) => {
                                                                                                        return method.runtime.NewText(Pair.GetNotice(on));
                                                                                                    })));
        }

        public override IokeData CloneData(IokeObject obj, IokeObject m, IokeObject context) {
            return new Pair(first, second);
        }


        public static string GetInspect(object on) {
            return ((Pair)(IokeObject.dataOf(on))).Inspect(on);
        }

        public static string GetNotice(object on) {
            return ((Pair)(IokeObject.dataOf(on))).Notice(on);
        }

        public override string ToString() {
            return "" + first + " => " + second;
        }

        public override string ToString(IokeObject obj) {
            return "" + first + " => " + second;
        }

        public string Inspect(object obj) {
            StringBuilder sb = new StringBuilder();

            sb.Append(IokeObject.Inspect(first));
            sb.Append(" => ");
            sb.Append(IokeObject.Inspect(second));

            return sb.ToString();
        }

        public string Notice(object obj) {
            StringBuilder sb = new StringBuilder();

            sb.Append(IokeObject.Notice(first));
            sb.Append(" => ");
            sb.Append(IokeObject.Notice(second));

            return sb.ToString();
        }
    }
}
